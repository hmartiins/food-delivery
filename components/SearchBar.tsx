import { useState } from 'react';

import { Image, TextInput, TouchableOpacity, View } from 'react-native';

import { router, useLocalSearchParams } from 'expo-router';

import { images } from '@/constants';

export const SearchBar = () => {
  const params = useLocalSearchParams<{ query: string }>();
  const [query, setQuery] = useState(params.query || '');

  const handleSearch = (text: string) => {
    setQuery(text);

    if (!text) router.setParams({ query: undefined });
  };

  const handleSubmit = () => {
    if (query?.trim()) router.setParams({ query });
  };

  return (
    <View className="searchbar" testID="search-container">
      <TextInput
        testID="search-input"
        className="flex-1 p-5"
        placeholder="Search for pizzas, burgers..."
        value={query}
        onChangeText={handleSearch}
        onSubmitEditing={handleSubmit}
        placeholderTextColor={'#a0a0a0'}
        returnKeyType="search"
      />

      <TouchableOpacity
        className="pr-5"
        onPress={() => router.setParams({ query })}
        testID="search-button"
      >
        <Image
          source={images.search}
          className="size-6"
          resizeMode="contain"
          tintColor={'#5d5f6d'}
        />
      </TouchableOpacity>
    </View>
  );
};
