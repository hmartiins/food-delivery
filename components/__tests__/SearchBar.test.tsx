import { router, useLocalSearchParams } from 'expo-router';

import { fireEvent, render } from '@testing-library/react-native';

import { SearchBar } from '../SearchBar';

// Mock do expo-router
jest.mock('expo-router', () => ({
  router: {
    setParams: jest.fn(),
  },
  useLocalSearchParams: jest.fn(),
}));

// Mock das constantes
jest.mock('@/constants', () => ({
  images: {
    search: 'search-icon',
  },
}));

const mockUseLocalSearchParams = useLocalSearchParams as jest.MockedFunction<
  typeof useLocalSearchParams
>;
describe('<SearchBar />', () => {
  beforeEach(() => jest.clearAllMocks());

  it('renders correctly with empty query', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByPlaceholderText, getByTestId } = render(<SearchBar />);

    expect(getByPlaceholderText('Search for pizzas, burgers...')).toBeTruthy();
    expect(getByTestId('search-button')).toBeTruthy();
  });

  it('renders correctly with initial query from params', () => {
    mockUseLocalSearchParams.mockReturnValue({ query: 'pizza' });

    const { getByDisplayValue } = render(<SearchBar />);

    expect(getByDisplayValue('pizza')).toBeTruthy();
  });

  it('updates query state when text input changes', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<SearchBar />);
    const textInput = getByTestId('search-input');

    fireEvent.changeText(textInput, 'burger');

    expect(textInput.props.value).toBe('burger');
  });

  it('calls router.setParams with undefined when text is cleared', () => {
    mockUseLocalSearchParams.mockReturnValue({ query: 'pizza' });

    const { getByDisplayValue } = render(<SearchBar />);
    const textInput = getByDisplayValue('pizza');

    fireEvent.changeText(textInput, '');

    expect(router.setParams).toHaveBeenCalledWith({ query: undefined });
  });

  it('calls router.setParams with query on submit editing when query has content', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<SearchBar />);
    const textInput = getByTestId('search-input');

    fireEvent.changeText(textInput, 'burger');
    fireEvent(textInput, 'submitEditing');

    expect(router.setParams).toHaveBeenCalledWith({ query: 'burger' });
  });

  it('does not call router.setParams on submit editing when query is empty', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<SearchBar />);
    const textInput = getByTestId('search-input');

    fireEvent(textInput, 'submitEditing');

    expect(router.setParams).not.toHaveBeenCalled();
  });

  it('does not call router.setParams on submit editing when query is only whitespace', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<SearchBar />);
    const textInput = getByTestId('search-input');

    fireEvent.changeText(textInput, '   ');
    fireEvent(textInput, 'submitEditing');

    expect(router.setParams).not.toHaveBeenCalled();
  });

  it('calls router.setParams when search button is pressed', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<SearchBar />);
    const textInput = getByTestId('search-input');
    const searchButton = getByTestId('search-button');

    fireEvent.changeText(textInput, 'pizza');
    fireEvent.press(searchButton);

    expect(router.setParams).toHaveBeenCalledWith({ query: 'pizza' });
  });

  it('calls router.setParams with empty string when search button is pressed with empty query', () => {
    mockUseLocalSearchParams.mockReturnValue({});

    const { getByTestId } = render(<SearchBar />);
    const searchButton = getByTestId('search-button');

    fireEvent.press(searchButton);

    expect(router.setParams).toHaveBeenCalledWith({ query: '' });
  });

  it('snapshot test', () => {
    mockUseLocalSearchParams.mockReturnValue({ query: 'test query' });

    const { toJSON } = render(<SearchBar />);
    expect(toJSON()).toMatchSnapshot();
  });
});
