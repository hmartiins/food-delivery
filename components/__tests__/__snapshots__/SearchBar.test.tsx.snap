// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> snapshot test 1`] = `
<View
  className="searchbar"
  testID="search-container"
>
  <TextInput
    className="flex-1 p-5"
    onChangeText={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Search for pizzas, burgers..."
    placeholderTextColor="#a0a0a0"
    returnKeyType="search"
    testID="search-input"
    value="test query"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="search-button"
  >
    <Image
      className="size-6"
      resizeMode="contain"
      source="search-icon"
      tintColor="#5d5f6d"
    />
  </View>
</View>
`;
